File: Python/DFT.py
# -*- coding: utf-8 -*-
"""
Created on Fri Apr 19 17:52:40 2024

@author: domin
"""
import numpy as np
from scipy import special
import scipy.linalg as sp
import matplotlib.pyplot as plt

from diis import DIIS
from molecular_integrals_s import molecule



class DFT():#molecule is actually integrals  
    """
    This is a class implementing KS-DFT code, it builds on molecular_integrals objects
    
    Density evaluation was written with the help of tutorials found in psi4numpy repository
    Psi4NumPy: An Interactive Quantum Chemistry Programming Environment for Reference Implementations and Rapid Development Daniel G. A. Smith, Lori A. Burns, Dominic A. Sirianni, Daniel R. Nascimento, Ashutosh Kumar, Andrew M. James, Jeffrey B. Schriber, Tianyuan Zhang, Boyi Zhang, Adam S. Abbott, Eric J. Berquist, Marvin H. Lechner, Leonardo A. Cunha, Alexander G. Heide, Jonathan M. Waldrop, Tyler Y. Takeshita, Asem Alenaizan, Daniel Neuhauser, Rollin A. King, Andrew C. Simmonett, Justin M. Turney, Henry F. Schaefer, Francesco A. Evangelista, A. Eugene DePrince III, T. Daniel Crawford, Konrad Patkowski, and C. David Sherrill Journal of Chemical Theory and Computation, 2018, 14 (7), 3504-3511 DOI: 10.1021/acs.jctc.8b00286
    """   
    def __init__(self,molecule,scf_param,functional="MP",limit=0):
        self.n_occ=molecule.n_occ
        self.get_grid(molecule)
        self.grid_atomic_orbitals(molecule.mol,self.grid)
        self.total_energy(scf_param,molecule,functional,limit)
    
    def get_grid(self,molecule):
        from grid import GRID 
        #there is also a grid2 script in which you can set coarse or close grid
        self.grid,self.weights=GRID(molecule)
    
    def grid_atomic_orbitals(self,basis,grid_point):
        ao=[]
        for gto in basis:
            psi=np.zeros(len(grid_point))
            for g in gto:
                psi+=gaussian_grid_eval(g,grid_point)
            ao.append(psi)
            
        self.ao=np.column_stack(ao)
        
        
    def evaluate_density(self):
        
        ao_density=np.einsum("pr,rq->pq", self.ao, self.D)
        ao_density=np.einsum("pi,pi->p", self.ao, ao_density)
        
        ao_density[abs(ao_density) < 1.0e-15] = 0
        
        self.rho=ao_density
        return ao_density
        
    def matrix_xc_potential(self,vxc):
        weighted_ao=np.einsum("pi,p->pi", self.ao, 0.5*self.weights*vxc)
        xc=np.einsum("rp,rq->pq", self.ao, weighted_ao)
        
        self.V_xc=xc+xc.T #to ensure symetry
        return xc+xc.T
    
    def xc_energy(self,exc):
        E_xc=np.einsum('p,p->', self.rho*self.weights, exc)
        self.E_xc=E_xc
        return E_xc
    
    def density_matrix(self):
        self.D=2*np.einsum("pi,qi->pq", self.C[:,:self.n_occ], self.C[:,:self.n_occ])
        
    def SCF(self,scf_param,molecule,functional,limit):
        import exco as ex
        convergence,diss_c, max_steps=scf_param
        ds=DIIS()
        
        e,self.C=sp.eigh(molecule.H_core,molecule.S)
        self.density_matrix()
        
        E0=0
        for step in range(max_steps):
            J=np.einsum("rs,pqrs->pq", self.D, molecule.V_ee)
            rho=self.evaluate_density()
            #exc,vxc=ex.Eexchange_LDA(rho)+ ex.Ecorrelation_LDA(ex.wigner_seitz_r(rho)), ex.Vexchange_LDA(rho)+ ex.Vcorrelation_LDA(ex.wigner_seitz_r(rho))
            exc,vxc=ex.Eexchange_LDA(rho)+ ex.e_correlation(ex.wigner_seitz_r(rho),functional,limit), ex.Vexchange_LDA(rho)+ ex.V_correlation(ex.wigner_seitz_r(rho),functional,limit)
           
            #V_xc=self.matrix_xc_potential(vxc) 
            #print(V_xc)
            fKS=molecule.H_core+J+self.matrix_xc_potential(vxc) 
            ds.update(fKS,self.D,molecule.S)
            
            if step>1:
                fKS=ds.DISS_F()
            
            e,self.C=sp.eigh(fKS,molecule.S)
            self.density_matrix()
            
            eKS=np.einsum('pq,pq->', self.D, (molecule.H_core + 1/2*J)) + self.xc_energy(exc)
            if (abs(eKS - E0) < convergence) and (ds.RMSD_check()<diss_c):
                #print("Converged to:", eKS)
                print("Converged")
                self.E_ee=eKS
                return
            E0=eKS
        print("Did not converge")
        
    def total_energy(self,scf_param,molecule,functional,limit):
        self.SCF(scf_param,molecule,functional,limit)
        self.E_total=self.E_ee+molecule.E_nn
        #print("Total energy is:", self.E_total)
    
def gaussian_grid_eval(gaussian,grid_point): 
    A=grid_point-gaussian.r
    psi=np.prod(A**gaussian.l, axis=1) * gaussian.N * gaussian.c * np.exp(-gaussian.a*np.sum(A*A, axis=1))
    return psi    

File: C++/DFT.cpp
#include <iostream>
#include <vector>
#include <eigen_lib/Eigen/Dense>
#include <eigen_lib/Eigen/Eigenvalues>
#include <boost/math/special_functions/gamma.hpp>
#include <integrals.h>
#include <diis.h>
#include <exco.h>
#include <grid.h>
#include <DFT.h>



 Eigen::VectorXd gaussian_grid_eval(Gaussian gaussian,std::vector<double> gridx,std::vector<double> gridy,std::vector<double> gridz){
    Eigen::VectorXd X(gridx.size());
    Eigen::VectorXd Y(gridy.size());
    Eigen::VectorXd Z(gridz.size());
    for (int i=0;i<gridx.size();++i)
    {
        X[i]=std::pow(gridx[i]-gaussian.r[0],gaussian.l[0]) *std::exp(-gaussian.a*std::pow(gridx[i]-gaussian.r[0],2.0));
        Y[i]=std::pow(gridy[i]-gaussian.r[1],gaussian.l[1]) *std::exp(-gaussian.a*std::pow(gridy[i]-gaussian.r[1],2.0));
        Z[i]=std::pow(gridz[i]-gaussian.r[2],gaussian.l[2]) *std::exp(-gaussian.a*std::pow(gridz[i]-gaussian.r[2],2.0));

    }
    Eigen::VectorXd phi(gridx.size());
    for (int j=0;j<gridx.size();++j)
    {
        phi[j]=X[j]*Y[j]*Z[j]*gaussian.N * gaussian.c;
    }
    return phi;
};
Eigen::MatrixXd get_J(Eigen::MatrixXd D, Eigen::Tensor<double,4> V_ee){
    Eigen::MatrixXd J(D.rows(),D.cols());
    J.setZero();
    auto dim =V_ee.dimensions();
    for (int p=0;p<dim[0];++p){
        for (int q=0;q<dim[1];++q){
            for (int r=0;r<dim[2];++r){
                for (int s=0;s<dim[3];++s){
                    J(p,q)+=D(r,s)*V_ee(p,q,r,s);
                }
            }
        }
    }
    return J;
}




DFT::DFT(MolecularIntegralsS molint, std::tuple<double, double, int> scf_param, int ty, std::string functional, int limit) {
        n_occ = molint.molecule.n_occ;
        get_grid(molint,ty);
        grid_atomic_orbitals(molint.molecule.mol, gridx, gridy, gridz);
        SCF(scf_param,molint,functional,limit);
        total_energy(molint);
    }
void DFT::get_grid(MolecularIntegralsS molint,int ty){
    auto tup=grid_maker(molint.molecule, ty);
    gridx=std::get<0>(tup);
    gridy=std::get<1>(tup);
    gridz=std::get<2>(tup);
    weights=std::get<3>(tup);
}

void DFT::grid_atomic_orbitals(std::vector<std::vector<Gaussian>> basis, std::vector<double> gridx,std::vector<double> gridy,std::vector<double> gridz){
    ao.resize(gridx.size(),basis.size());
    for (int i=0;i<basis.size();++i){
        auto gto=basis[i];
        Eigen::VectorXd psi(gridx.size());
        psi.setZero();
        for (int j=0;j<gto.size();++j){
            auto g=gto[j];
            psi+=gaussian_grid_eval(g,gridx,gridy,gridz);
        ao.col(i)=psi;

        }
    }
}

Eigen::VectorXd DFT::evaluate_density(){
    Eigen::VectorXd aoden(gridx.size());
    aoden.setZero();
    
    for (int p=0;p<aoden.size();++p){
        for (int u=0;u<D.rows();++u){
            for (int v=0;v<D.cols();++v){
                    aoden[p]+=ao(p,u)*D(u,v)*ao(p,v);
                    
            }
        }
    }
    
    rho=aoden;
    return aoden;
}

Eigen::MatrixXd DFT::matrix_xc_potential(Eigen::VectorXd vxc){
    Eigen::MatrixXd V_temp(ao.cols(),ao.cols());
    V_temp.setZero();
    for (int p=0;p<ao.rows();++p){
        for (int u=0;u<ao.cols();++u){
            for (int v=0;v<ao.cols();++v){
                    V_temp(u,v)+=ao(p,u)*0.5*weights[p]*vxc[p]*ao(p,v);
            }
        }
    }
    V_xc=V_temp+V_temp.transpose();
    return V_temp+V_temp.transpose();
}

double DFT::xc_energy(Eigen::VectorXd exc){
    double xc=0;
    for (int p=0;p<exc.size();++p){
        xc+=rho[p]*weights[p]*exc[p];
    }
    E_xc=xc;
    return xc;
}

void DFT::density_matrix() {
        D.resize(C.rows(), C.rows());
        D.setZero();
        Eigen::MatrixXd C1=C(Eigen::all,Eigen::seq(0,n_occ-1));

        for (int p=0; p<C1.rows(); ++p){
            for (int q=0; q<C1.rows(); ++q){
                for (int i=0; i<C1.cols(); ++i)
                {
                    
                    D(p,q)+=2*C1(p,i)*C1(q,i);
                }
            }   
        }
    }

void DFT::SCF(std::tuple<double, double, int> scf_param,MolecularIntegralsS molint, std::string functional, int limit){
    double convergence=std::get<0>(scf_param);
    double diss_c=std::get<1>(scf_param);
    int max_steps=std::get<2>(scf_param);
    diis ds;
    Eigen::GeneralizedSelfAdjointEigenSolver<Eigen::MatrixXd> es(molint.H_core,molint.S);
    C=es.eigenvectors();
    density_matrix();
    
    double E0=0;
    for (int step = 0; step < max_steps; ++step) 
    {
        Eigen::MatrixXd   J=get_J(D,molint.V_ee);
        
        Eigen::VectorXd rho=evaluate_density();
        double norm=0;
        for (int i=0;i<rho.size();++i){norm+=rho[i]*weights[i];}

        Eigen::VectorXd vxc=Vexchange_LDA(rho) + V_correlation(wigner_seitz_r(rho),functional,limit); //can be changed to WVN3
        Eigen::VectorXd exc=Eexchange_LDA(rho) + E_correlation(wigner_seitz_r(rho),functional,limit);

        Eigen::MatrixXd  fKS=molint.H_core+J+matrix_xc_potential(vxc);
       
        ds.update(fKS,D,molint.S);
        if (step>1){
            fKS=ds.DIIS_F();
        
        }
        
        Eigen::GeneralizedSelfAdjointEigenSolver<Eigen::MatrixXd> es(fKS,molint.S);
        C=es.eigenvectors();
        density_matrix();

        Eks=0;
        for (int p=0;p<D.rows();++p){
            for (int q=0;q<D.cols();++q){
                Eks+=D(p,q)*(molint.H_core(p,q)+0.5*J(p,q));
            }
        }
        Eks+=xc_energy(exc);

        std::cout <<std::setprecision(17)<< Eks << std::endl;
        double rd=ds.RMSD_check();
       

        if ((abs(Eks - E0) < convergence) && (rd<diss_c)) {
            std::cout << "Converged\n";
            E_ee = Eks;
            return;
        }
        E0 = Eks;
    }
    std::cout << "Did not converge\n";
}


void DFT::total_energy(MolecularIntegralsS molint){
    E_total=E_ee+molint.E_nn;
}
